### PR TITLE
[IMP] stock: archive warehouse

### PR DESCRIPTION
**Pull request by amoyaux**
[IMP] stock: archive warehouse

Archiving a warehouse didn't archive its related locations, routes and
rules.

So we now archive the related objects and add some constraints to avoid
having an inconsistent state in stock with archived warehouse.

TASK-ID: 30659

### CODE REVIEW

File: addons/stock/models/stock_warehouse.py [hunk 1]
```diff
@@ -171,13 +171,118 @@ def write(self, vals):
                 to_add = new_resupply_whs - old_resupply_whs[warehouse.id]
                 to_remove = old_resupply_whs[warehouse.id] - new_resupply_whs
                 if to_add:
-                    warehouse.create_resupply_routes(to_add)
+                    exisiting_route = Route.search([
```
Comment (mreficent) [inline addons/stock/models/stock_warehouse.py]: existing_route

File: addons/stock/models/stock_warehouse.py [hunk 2]
```diff
@@ -171,13 +171,116 @@ def write(self, vals):
...
+            move_ids = self.env['stock.move'].search([
+                ('picking_type_id', 'in', picking_type_ids.ids)
+            ])
+            if move_ids:
+                raise UserError('You still have ongoing operations for picking\
+                    types %s in warehouse %s' %
+                    (', '.join(move_ids.mapped('picking_type_id.name')), warehouse.name))
+
+            location_ids = warehouse._get_warehouse_locations(active=warehouse.active)
+            quant_ids = self.env['stock.quant'].search([
```
Comment (pimodoo) [inline addons/stock/models/stock_warehouse.py]: move me and my bros in toggle active of locations

File: addons/stock/models/stock_warehouse.py [hunk 3]
```diff
@@ -166,12 +166,68 @@ def write(self, vals):
                     any(depend in vals for depend in depends):
                 warehouse._create_or_update_global_routes_rules()
 
+            if 'active' in vals:
+                picking_type_ids = self.env['stock.picking.type'].with_context(active_test=False).search([('warehouse_id', '=', warehouse.id)])
+                move_ids = self.env['stock.move'].search([
+                    ('picking_type_id', 'in', picking_type_ids.ids)
```
Comment (sle-odoo) [inline addons/stock/models/stock_warehouse.py]: check the state

File: addons/stock/models/stock_warehouse.py [hunk 4]
```diff
@@ -166,12 +166,68 @@ def write(self, vals):
...
+                ])
+                if move_ids:
+                    raise UserError('You still have ongoing operations for picking\
```
Comment (sle-odoo) [inline addons/stock/models/stock_warehouse.py]: translation?

File: addons/stock/models/stock_location.py [hunk 5]
```diff
@@ -97,18 +97,15 @@ def write(self, values):
...
-            if quant_ids:
-                raise UserError(_('You still have some product in locations %s') %
-                    (quant_ids.mapped('location_id.name')))
+            if not self.env.context.get('do_not_check_quant'):
+                children_location = self.env['stock.location'].with_context(active_test=False).search([('id', 'child_of', self.ids)])
+                internal_children_locations = children_location.filtered(lambda l: l.usage == 'internal')
+                children_quants = self.env['stock.quant'].search([('quantity', '!=', 0), ('reserved_quantity', '!=', 0), ('location_id', 'in', internal_children_locations.ids)])
+                if children_quants and values['active'] == False:
+                    raise UserError(_('You still have some product in locations %s') %
+                        (children_quants.mapped('location_id.name')))
```
Comment (sle-odoo) [inline addons/stock/models/stock_location.py]: maybe display_name?
join the locations with a comma

File: addons/stock/models/stock_warehouse.py [hunk 6]
```diff
@@ -166,12 +166,69 @@ def write(self, vals):
...
+                        types %s in warehouse %s') %
+                        (', '.join(move_ids.mapped('picking_type_id.name')), warehouse.name))
+                else:
+                    picking_type_ids.write({'active': vals['active']})
+                location_ids = self.env['stock.location'].with_context(active_test=False).search([('location_id', 'child_of', warehouse.view_location_id.id)])
+                picking_type_using_locations = self.env['stock.picking.type'].search([
+                    ('default_location_src_id', 'in', location_ids.ids),
+                    ('default_location_dest_id', 'in', location_ids.ids),
+                    ('id', 'not in', picking_type_ids.ids),
+                    ('warehouse_id', '=', warehouse.id),
```
Comment (sle-odoo) [inline addons/stock/models/stock_warehouse.py]: are you sure this line makes sense? what are you trying to achieve?

File: addons/stock/models/stock_warehouse.py [hunk 7]
```diff
@@ -166,12 +166,69 @@ def write(self, vals):
...
+                    raise UserError(_('%s use default source or destination locations\
+                        from warehouse %s that will be archived.') %
+                        (', '.join(picking_type_using_locations.mapped('name')), warehouse.name))
+                warehouse.view_location_id.write({'active': vals['active']})
+
+                rule_ids = self.env['stock.rule'].with_context(active_test=False).search([('warehouse_id', '=', self.id)])
+                # Only modify route that apply on this warehouse.
+                route_ids = warehouse.route_ids.filtered(lambda r: len(r.warehouse_ids) == 1)
+                # Active/archive all warehouse components.
+                route_ids.filtered(lambda r: not r.warehouse_ids - warehouse).write({'active': vals['active']})
```
Comment (sle-odoo) [inline addons/stock/models/stock_warehouse.py]: this filtered is useless i think

Comment (pimodoo) [reply addons/stock/models/stock_warehouse.py]: nope, it is to avoid removing routes linked to multiple warehouses.

Comment (sle-odoo) [reply addons/stock/models/stock_warehouse.py]: but you already filter on route having a single warehouse_ids in the previous line

Comment (pimodoo) [reply addons/stock/models/stock_warehouse.py]: true

**Comment by sle-odoo** [comment]
robodoo r+

**Pull request closed without merging**